File: src/parser/tokenizer.ts
/**
 * Tokenizer for Arkitecture DSL
 */

export enum TokenType {
  IDENTIFIER = 'IDENTIFIER',
  STRING = 'STRING',
  NUMBER = 'NUMBER',
  LBRACE = 'LBRACE', // {
  RBRACE = 'RBRACE', // }
  LBRACKET = 'LBRACKET', // [
  RBRACKET = 'RBRACKET', // ]
  COLON = 'COLON', // :
  COMMA = 'COMMA', // ,
  ARROW = 'ARROW', // -->
  DOT = 'DOT', // .
  HASH = 'HASH', // #
  GROUP = 'GROUP', // group keyword
  EOF = 'EOF',
  NEWLINE = 'NEWLINE',
}

export interface Token {
  type: TokenType;
  value: string;
  line: number;
  column: number;
}

export class TokenizerError extends Error {
  constructor(
    message: string,
    public line: number,
    public column: number
  ) {
    super(`${message} at line ${line}, column ${column}`);
    this.name = 'TokenizerError';
  }
}

export class Tokenizer {
  private input: string;
  private position: number;
  private line: number;
  private column: number;

  constructor(input: string) {
    this.input = input;
    this.position = 0;
    this.line = 1;
    this.column = 1;
  }

  tokenize(): Token[] {
    const tokens: Token[] = [];
    
    while (!this.isAtEnd()) {
      const token = this.nextToken();
      if (token) {
        tokens.push(token);
      }
    }
    
    tokens.push({
      type: TokenType.EOF,
      value: '',
      line: this.line,
      column: this.column,
    });
    
    return tokens;
  }

  private nextToken(): Token | null {
    this.skipWhitespace();
    
    if (this.isAtEnd()) {
      return null;
    }
    
    const startLine = this.line;
    const startColumn = this.column;
    const char = this.peek();
    
    // Comments vs Hash character
    if (char === '#') {
      // If the # is preceded by whitespace or start of line, it's a comment
      // Otherwise, it's a hash token for anchor references
      if (this.column === 1 || this.isPrecedingWhitespace()) {
        this.skipComment();
        return this.nextToken();
      } else {
        this.advance();
        return {
          type: TokenType.HASH,
          value: '#',
          line: startLine,
          column: startColumn,
        };
      }
    }
    
    // Newlines
    if (char === '\n') {
      this.advance();
      return {
        type: TokenType.NEWLINE,
        value: '\n',
        line: startLine,
        column: startColumn,
      };
    }
    
    // Arrow operator
    if (char === '-' && this.peek(1) === '-' && this.peek(2) === '>') {
      this.advance();
      this.advance();
      this.advance();
      return {
        type: TokenType.ARROW,
        value: '-->',
        line: startLine,
        column: startColumn,
      };
    }
    
    // Single character tokens
    switch (char) {
      case '{':
        this.advance();
        return {
          type: TokenType.LBRACE,
          value: '{',
          line: startLine,
          column: startColumn,
        };
      case '}':
        this.advance();
        return {
          type: TokenType.RBRACE,
          value: '}',
          line: startLine,
          column: startColumn,
        };
      case '[':
        this.advance();
        return {
          type: TokenType.LBRACKET,
          value: '[',
          line: startLine,
          column: startColumn,
        };
      case ']':
        this.advance();
        return {
          type: TokenType.RBRACKET,
          value: ']',
          line: startLine,
          column: startColumn,
        };
      case ':':
        this.advance();
        return {
          type: TokenType.COLON,
          value: ':',
          line: startLine,
          column: startColumn,
        };
      case ',':
        this.advance();
        return {
          type: TokenType.COMMA,
          value: ',',
          line: startLine,
          column: startColumn,
        };
      case '.':
        this.advance();
        return {
          type: TokenType.DOT,
          value: '.',
          line: startLine,
          column: startColumn,
        };
    }
    
    // String literals
    if (char === '"') {
      return this.scanString(startLine, startColumn);
    }
    
    // Numbers
    if (this.isDigit(char)) {
      return this.scanNumber(startLine, startColumn);
    }
    
    // Identifiers and keywords
    if (this.isAlpha(char)) {
      return this.scanIdentifier(startLine, startColumn);
    }
    
    throw new TokenizerError(
      `Unexpected character '${char}'`,
      startLine,
      startColumn
    );
  }

  private scanString(startLine: number, startColumn: number): Token {
    this.advance(); // Skip opening quote
    const value: string[] = [];
    
    while (!this.isAtEnd() && this.peek() !== '"') {
      if (this.peek() === '\n') {
        this.advance();
        this.line++;
        this.column = 1;
        value.push('\n');
      } else if (this.peek() === '\\') {
        this.advance(); // Skip backslash
        if (this.isAtEnd()) {
          throw new TokenizerError(
            'Unterminated string escape',
            startLine,
            startColumn
          );
        }
        const escaped = this.advance();
        switch (escaped) {
          case 'n':
            value.push('\n');
            break;
          case 't':
            value.push('\t');
            break;
          case 'r':
            value.push('\r');
            break;
          case '\\':
            value.push('\\');
            break;
          case '"':
            value.push('"');
            break;
          default:
            value.push(escaped);
        }
      } else {
        value.push(this.advance());
      }
    }
    
    if (this.isAtEnd()) {
      throw new TokenizerError(
        'Unterminated string',
        startLine,
        startColumn
      );
    }
    
    this.advance(); // Skip closing quote
    
    return {
      type: TokenType.STRING,
      value: value.join(''),
      line: startLine,
      column: startColumn,
    };
  }

  private scanNumber(startLine: number, startColumn: number): Token {
    const value: string[] = [];
    
    while (this.isDigit(this.peek())) {
      value.push(this.advance());
    }
    
    // Handle decimal numbers
    if (this.peek() === '.' && this.isDigit(this.peek(1))) {
      value.push(this.advance()); // Add the '.'
      while (this.isDigit(this.peek())) {
        value.push(this.advance());
      }
    }
    
    return {
      type: TokenType.NUMBER,
      value: value.join(''),
      line: startLine,
      column: startColumn,
    };
  }

  private scanIdentifier(startLine: number, startColumn: number): Token {
    const value: string[] = [];
    
    while (this.isAlphaNumeric(this.peek()) || this.peek() === '_') {
      value.push(this.advance());
    }
    
    const text = value.join('');
    const type = text === 'group' ? TokenType.GROUP : TokenType.IDENTIFIER;
    
    return {
      type,
      value: text,
      line: startLine,
      column: startColumn,
    };
  }

  private skipWhitespace(): void {
    while (!this.isAtEnd()) {
      const char = this.peek();
      if (char === ' ' || char === '\r' || char === '\t') {
        this.advance();
      } else {
        break;
      }
    }
  }

  private skipComment(): void {
    while (!this.isAtEnd() && this.peek() !== '\n') {
      this.advance();
    }
  }

  private peek(offset: number = 0): string {
    const pos = this.position + offset;
    if (pos >= this.input.length) {
      return '\\0';
    }
    return this.input[pos];
  }

  private advance(): string {
    if (this.isAtEnd()) {
      return '\\0';
    }
    
    const char = this.input[this.position];
    this.position++;
    
    if (char === '\n') {
      this.line++;
      this.column = 1;
    } else {
      this.column++;
    }
    
    return char;
  }

  private isAtEnd(): boolean {
    return this.position >= this.input.length;
  }

  private isDigit(char: string): boolean {
    return char >= '0' && char <= '9';
  }

  private isAlpha(char: string): boolean {
    return (char >= 'a' && char <= 'z') ||
           (char >= 'A' && char <= 'Z') ||
           char === '_';
  }

  private isAlphaNumeric(char: string): boolean {
    return this.isAlpha(char) || this.isDigit(char);
  }

  private isPrecedingWhitespace(): boolean {
    if (this.position === 0) {
      return true;
    }
    const prevChar = this.input[this.position - 1];
    return prevChar === ' ' || prevChar === '\t' || prevChar === '\r' || prevChar === '\n';
  }
}

File: parser/tokenizer.go
package parser

import (
	"fmt"
	"strings"
)

// TokenizerError is a lexical error with its source position.
type TokenizerError struct {
	Message string
	Line    int
	Column  int
}

func (e *TokenizerError) Error() string {
	return fmt.Sprintf("%s at line %d, column %d", e.Message, e.Line, e.Column)
}

// tokenizer scans DSL source over a slice of runes, tracking 1-based line and
// column positions.
type tokenizer struct {
	input    []rune
	position int
	line     int
	column   int
}

// Tokenize scans input into a slice of tokens terminated by a TokenEOF. It
// returns a *TokenizerError on the first lexical error (an unexpected character
// or an unterminated string).
func Tokenize(input string) ([]Token, *TokenizerError) {
	t := &tokenizer{input: []rune(input), position: 0, line: 1, column: 1}
	var tokens []Token
	for !t.isAtEnd() {
		tok, err := t.nextToken()
		if err != nil {
			return nil, err
		}
		if tok != nil {
			tokens = append(tokens, *tok)
		}
	}
	tokens = append(tokens, Token{Type: TokenEOF, Value: "", Line: t.line, Column: t.column})
	return tokens, nil
}

// nextToken returns the next significant token, skipping whitespace and
// comments. It returns (nil, nil) when only trailing whitespace/comments remain.
func (t *tokenizer) nextToken() (*Token, *TokenizerError) {
	for {
		t.skipWhitespace()
		if t.isAtEnd() {
			return nil, nil
		}

		startLine := t.line
		startColumn := t.column
		ch := t.peek(0)

		// '#' is overloaded three ways. Glued to an identifier (`node#anchor`) it
		// is an anchor hash. Otherwise it introduces either a hex colour value
		// (`#ff0000`) or a comment — a complete hex run with a word boundary after
		// it is a colour; anything else (a space, a non-hex word) is a comment.
		if ch == '#' {
			if t.isPrecedingIdentChar() {
				t.advance()
				return &Token{Type: TokenHash, Value: "#", Line: startLine, Column: startColumn}, nil
			}
			if tok := t.scanColor(startLine, startColumn); tok != nil {
				return tok, nil
			}
			t.skipComment()
			continue
		}

		if ch == '\n' {
			t.advance()
			return &Token{Type: TokenNewline, Value: "\n", Line: startLine, Column: startColumn}, nil
		}

		if ch == '@' {
			t.advance()
			return &Token{Type: TokenAt, Value: "@", Line: startLine, Column: startColumn}, nil
		}

		if ch == '-' && t.peek(1) == '-' && t.peek(2) == '>' {
			t.advance()
			t.advance()
			t.advance()
			return &Token{Type: TokenArrow, Value: "-->", Line: startLine, Column: startColumn}, nil
		}

		switch ch {
		case '{':
			t.advance()
			return &Token{Type: TokenLBrace, Value: "{", Line: startLine, Column: startColumn}, nil
		case '}':
			t.advance()
			return &Token{Type: TokenRBrace, Value: "}", Line: startLine, Column: startColumn}, nil
		case '[':
			t.advance()
			return &Token{Type: TokenLBracket, Value: "[", Line: startLine, Column: startColumn}, nil
		case ']':
			t.advance()
			return &Token{Type: TokenRBracket, Value: "]", Line: startLine, Column: startColumn}, nil
		case ':':
			t.advance()
			return &Token{Type: TokenColon, Value: ":", Line: startLine, Column: startColumn}, nil
		case ',':
			t.advance()
			return &Token{Type: TokenComma, Value: ",", Line: startLine, Column: startColumn}, nil
		case '.':
			t.advance()
			return &Token{Type: TokenDot, Value: ".", Line: startLine, Column: startColumn}, nil
		}

		if ch == '"' {
			return t.scanString(startLine, startColumn)
		}
		if isDigit(ch) {
			return t.scanNumber(startLine, startColumn), nil
		}
		if isAlpha(ch) {
			return t.scanIdentifier(startLine, startColumn), nil
		}

		return nil, &TokenizerError{
			Message: fmt.Sprintf("Unexpected character '%c'", ch),
			Line:    startLine,
			Column:  startColumn,
		}
	}
}

func (t *tokenizer) scanString(startLine, startColumn int) (*Token, *TokenizerError) {
	t.advance() // opening quote
	var b strings.Builder

	for !t.isAtEnd() && t.peek(0) != '"' {
		switch t.peek(0) {
		case '\n':
			t.advance() // advance() tracks the line/column change
			b.WriteRune('\n')
		case '\\':
			t.advance() // consume backslash
			if t.isAtEnd() {
				return nil, &TokenizerError{Message: "Unterminated string escape", Line: startLine, Column: startColumn}
			}
			switch esc := t.advance(); esc {
			case 'n':
				b.WriteRune('\n')
			case 't':
				b.WriteRune('\t')
			case 'r':
				b.WriteRune('\r')
			case '\\':
				b.WriteRune('\\')
			case '"':
				b.WriteRune('"')
			default:
				b.WriteRune(esc)
			}
		default:
			b.WriteRune(t.advance())
		}
	}

	if t.isAtEnd() {
		return nil, &TokenizerError{Message: "Unterminated string", Line: startLine, Column: startColumn}
	}
	t.advance() // closing quote
	return &Token{Type: TokenString, Value: b.String(), Line: startLine, Column: startColumn}, nil
}

func (t *tokenizer) scanNumber(startLine, startColumn int) *Token {
	var b strings.Builder
	for isDigit(t.peek(0)) {
		b.WriteRune(t.advance())
	}
	if t.peek(0) == '.' && isDigit(t.peek(1)) {
		b.WriteRune(t.advance()) // the '.'
		for isDigit(t.peek(0)) {
			b.WriteRune(t.advance())
		}
	}
	return &Token{Type: TokenNumber, Value: b.String(), Line: startLine, Column: startColumn}
}

// scanColor reads a hex colour token starting at the '#'. It returns nil (so the
// caller falls back to comment handling) unless the '#' is followed by a run of
// hex digits ending at a word boundary — e.g. `#ff0000`, `#abc`. A run trailed by
// a non-hex word char (`#fffg`) or no run at all (`# note`) is not a clean colour,
// so it stays a comment. The exact digit count (3/4/6/8) is the validator's check;
// here any boundary-terminated hex run becomes a TokenColor so a wrong length can
// be reported with a helpful message rather than silently swallowed as a comment.
func (t *tokenizer) scanColor(startLine, startColumn int) *Token {
	n := 1
	for isHexDigit(t.peek(n)) {
		n++
	}
	if n == 1 {
		return nil // '#' not followed by any hex digit
	}
	if isAlphaNumeric(t.peek(n)) {
		return nil // a non-hex word char follows the run: not a clean colour
	}
	var b strings.Builder
	for i := 0; i < n; i++ {
		b.WriteRune(t.advance())
	}
	return &Token{Type: TokenColor, Value: b.String(), Line: startLine, Column: startColumn}
}

func (t *tokenizer) scanIdentifier(startLine, startColumn int) *Token {
	var b strings.Builder
	for isAlphaNumeric(t.peek(0)) || t.peek(0) == '_' {
		b.WriteRune(t.advance())
	}
	return &Token{Type: TokenIdentifier, Value: b.String(), Line: startLine, Column: startColumn}
}

func (t *tokenizer) skipWhitespace() {
	// ';' is a purely cosmetic statement separator (it lets `@layout {a; b}`
	// read on one line), so it is skipped like ordinary whitespace.
	for !t.isAtEnd() {
		switch t.peek(0) {
		case ' ', '\r', '\t', ';':
			t.advance()
		default:
			return
		}
	}
}

func (t *tokenizer) skipComment() {
	for !t.isAtEnd() && t.peek(0) != '\n' {
		t.advance()
	}
}

// peek returns the rune at the current position plus offset, or 0 past the end.
func (t *tokenizer) peek(offset int) rune {
	pos := t.position + offset
	if pos < 0 || pos >= len(t.input) {
		return 0
	}
	return t.input[pos]
}

func (t *tokenizer) advance() rune {
	if t.isAtEnd() {
		return 0
	}
	ch := t.input[t.position]
	t.position++
	if ch == '\n' {
		t.line++
		t.column = 1
	} else {
		t.column++
	}
	return ch
}

func (t *tokenizer) isAtEnd() bool {
	return t.position >= len(t.input)
}

func (t *tokenizer) isPrecedingWhitespace() bool {
	if t.position == 0 {
		return true
	}
	switch t.input[t.position-1] {
	case ' ', '\t', '\r', '\n':
		return true
	default:
		return false
	}
}

// isPrecedingIdentChar reports whether the rune just before the cursor is an
// identifier char — the test that distinguishes an anchor hash (`node#a`, glued
// to an identifier) from a colour/comment `#` (preceded by space, ':', '{', …).
func (t *tokenizer) isPrecedingIdentChar() bool {
	return t.position > 0 && isAlphaNumeric(t.input[t.position-1])
}

func isDigit(c rune) bool { return c >= '0' && c <= '9' }

func isHexDigit(c rune) bool {
	return isDigit(c) || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F')
}

func isAlpha(c rune) bool {
	return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '_'
}

func isAlphaNumeric(c rune) bool { return isAlpha(c) || isDigit(c) }
